pkg/apis/apps/types.go

```diff
@@ -863,6 +863,9 @@ const (
 	// due to insufficient quota, limit ranges, pod security policy, node selectors, etc. or deleted
 	// due to kubelet being down or finalizers are failing.
 	ReplicaSetReplicaFailure ReplicaSetConditionType = "ReplicaFailure"
+	// ReplicaSetAvailable means the replica set is available, ie. all required replicas are up
+	// and running for at least minReadySeconds.
+	ReplicaSetAvailable ReplicaSetConditionType = "Available"
```

Comment:
... is this really a good condition to have? I imagine it will almost always be false for sufficiently large replica sets. 100% ready is an unrealistic goal, increasingly so for bigger sets.

Reply:
We are trying to consolidate workloads conditions as mentioned in the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/2804-consolidate-workload-controllers-status). The goal is to have similar interface across the workloads to signal to the users how their workload is performing.

I think it should not be a desired state of an application to  struggle to bring some of its replicas to ready/available status. And even in cases where some applications are intentionally overcommitted this would help to analyze such cases and give clear signal if all of the replicas are `Available` or not.

Also the ReplicaSet is underlying implementation of a Deployment (which in some configurations behaves almost like a ReplicaSet) and it already has `Available` condition  so users are already familiar with this concept.

Reply:
> Sorry for slow response, I missed this in my email

Np at all.

> "available" is not a good word. It's binary and ambiguous. I personally would expect that a service with 90% of its pods running is very likely available and totally fine, but that's clearly not your interpretation. (I can even imagine a service with 100% of its pods up being unavailable because a broken network is not sending them any traffic!)

> But also let me expand on my original comment. To be useful, we want a condition to be in the good state most of the time in most probable uses. The anomalous state should be rare, otherwise people will learn to ignore the condition. Yet, for a readily imaginable use case (large, auto-scaled deployment) this condition will probably generally be false. This indicates it is not a useful condition. I do not think we want to train operators of large services that the "available" condition should be ignored!

I see your point, the main problem stems from the fact that ReplicaSet is made simple and does not have the rich functionality of a Deployment (intentionally) so the user cannot define what availability means for them through eg `maxUnavailable` as in the Deployment and would be forced to accept either 100% availability or nothing.

> If we want to communicate how many of the replicas are ready, a count is a better way to do that than a condition.

we already do that for both readyReplicas and availableReplicas

```
status:
    availableReplicas: 1
    fullyLabeledReplicas: 1
    observedGeneration: 1
    readyReplicas: 1
    replicas: 1
```

> you either need to find a word that means the thing you mean ("100% of pods are ready and have been for a while" -- which I maintain that few people want),

I suppose we could rename it something like `AllReplicasAvailable`, but I have to note that one of the goals of the KEP was to have similarly named conditions if possible accross all the workloads so they would be familiar to the users.

> or you need to change the way this will be computed to mean something more like "the service as a whole is healthy and functioning" -- which does not imply that every single pod is up and hasn't been restarted recently.

This is probably a task for other controllers/components. I think the farthest we want to go is to utilize `minReadySeconds` / `availableReplicas` .

I agree with negatives, but still think it could be useful, but can drop this or just keep it open if necessary. And take on the other conditions from the KEP.

Reply:
@ravisantoshgudimetla @soltysh  can I please also get your opinions on this? ^

Reply:
Given all that, I think I'd advocate for not including an "available" condition on replicasets at all.

Especially since you can get this information already by looking at the counters in status.

I get the goal of putting common conditions on things, but if the condition isn't really meaningful for the object, or has a surprising meaning, I don't think we're actually doing our users a favor.

Reply:
@atiratree after reading Daniel's comment and looking through dictionary I'm leaning towards replacing `Available` with something like `Functional` which to me seems like a reasonable alternative expressing the desire in a clean way w/o introducing the necessary problem. @lavalamp wdyt?

Btw. `maxUnavilable` in deployment isn't meant for expressing availability in general, but only during rolling out a new revision of the application. In most cases the desire is to reach 100%, although the cases that Daniel explains seem reasonable, I'll probably need to dig through some data, I'm very curious about that approach.

Reply:
`Functional` also doesn't seem right, since you don't actually know that something with 90% of its replicas healthy is functional, nor that one with 10% healthy isn't.

`NominallyAvailable` is maybe something you could use to mean "enough of this thing is available that you're probably good".